src/policy/compressor/forwarding.rs

```diff
@@ -0,0 +1,165 @@
+
+/// A finite-state machine which processes the positions of mark bits,
+/// and accumulates the size of live data that it has seen.
+///
+/// The Compressor caches the state of the transducer at the start of
+/// each block by serialising the state using [`Transducer::encode`];
+/// the state can then be deserialised using [`Transducer::decode`].
+#[derive(Debug)]
+struct Transducer {
+    live: usize,
+    last_bit_seen: Address,
+    in_object: bool,
+}
+impl Transducer {
+    pub fn new() -> Self {
+        Self {
+            live: 0,
+            last_bit_seen: Address::ZERO,
+            in_object: false,
+        }
+    }
+    pub fn step(&mut self, address: Address) {
+        if self.in_object {
+            self.live += address - self.last_bit_seen + MIN_OBJECT_SIZE;
+        }
+        self.in_object = !self.in_object;
+        self.last_bit_seen = address;
+    }
+
+    pub fn encode(&self, address: Address) -> usize {
+        if self.in_object {
+            // We count the space between the last mark bit and
+            // the current address as live when we stop in the
+            // middle of an object.
+            self.live + (address - self.last_bit_seen) + 1
+        } else {
+            self.live
+        }
+    }
+
+    pub fn decode(offset: usize, address: Address) -> Self {
+        Transducer {
+            live: offset & !1,
+            last_bit_seen: address,
+            in_object: (offset & 1) == 1,
+        }
+    }
+}
+
+pub struct ForwardingMetadata<VM: VMBinding> {
+    pub(crate) first_address: Address,
+    calculated: AtomicBool,
+    vm: PhantomData<VM>,
+}
+
+// A block in the Compressor is the granularity at which we record
+// the live data prior to the start of each block. We set it to 512 bytes
+// following the paper.
+pub(crate) const LOG_BLOCK_SIZE: usize = 9;
+pub(crate) const BLOCK_SIZE: usize = 1 << LOG_BLOCK_SIZE;
+pub(crate) const MARK_SPEC: SideMetadataSpec = COMPRESSOR_MARK;
+pub(crate) const OFFSET_VECTOR_SPEC: SideMetadataSpec = COMPRESSOR_OFFSET_VECTOR;
+
+impl<VM: VMBinding> ForwardingMetadata<VM> {
+    pub fn new(start: Address) -> ForwardingMetadata<VM> {
+        ForwardingMetadata {
+            first_address: start,
+            calculated: AtomicBool::new(false),
+            vm: PhantomData,
+        }
+    }
+
+    pub fn mark_end_of_object(&self, object: ObjectReference) {
+        let end_of_object = object.to_object_start::<VM>()
+            + VM::VMObjectModel::get_current_size(object)
+            - MIN_OBJECT_SIZE;
+        #[cfg(debug_assertions)]
+        {
+            use crate::util::metadata::side_metadata::{address_to_meta_address, meta_byte_lshift};
+            // We require to be able to iterate upon start and end bits in the
+            // same bitmap. Therefore the start and end bits cannot be the
+            // same, else we would only encounter one of the two bits.
+            let a1 = address_to_meta_address(&MARK_SPEC, object.to_raw_address());
```

Comment:
Could add a function to `SideMetadataSpec`, such as `fn are_separate_bits(&self, addr1, addr2) -> bool`. In that case, we don't need to expose `meta_byte_lshift` as `pub(crate)`.